sklearn/tests/test_base.py

```diff
@@ -178,7 +178,7 @@ def test_repr():
     )
 
     some_est = T(a=["long_params"] * 1000)
-    assert_equal(len(repr(some_est)), 415)
+    assert_equal(len(repr(some_est)), 16015)
```

Comment:
I think the point of this test was to check that the repr would not be excessively long, and that ellipsis would be used. I'm not convinced that this change is faithful to that concern

Reply:
You're correct.
We're not using ellipsis anymore, should we add it?

I doubt it would be really useful now that the pprint is correctly formatted, supposing that we would almost never have extremely long reprs.

Reply:
We do by default, only with the short repr we wouldn't. Maybe we could switch to the short repr if the full repr is too long? That seems more useful than ellipsis?

Reply:
What do you mean by the short repr? This test implies that the default repr is much longer than the current default

Reply:
by short I meant with ``changed_only=True``

Reply:
What if it's a DictVectorizer with a provided vocabulary?

Reply:
I have no idea what the repr for this is in this pr. @NicolasHug ?
We might want to limit some things in length more than others maybe?

Reply:
That is, we should bound use ellipsis for non-estimators

Reply:
What do you mean a DictVectorizer with provided vocabulary?

`__repr__` (in this PR and in the current implem) only cares about the arguments passed to `__init__`, not those to `fit`, so the `vocabulary_` attribute would be ignored:

```python
from sklearn.feature_extraction import DictVectorizer
v = DictVectorizer(sparse=False)
D = [{'foo': 1, 'bar': 2}, {'foo': 3, 'baz': 1}]
v.fit(D)
v
```

```
DictVectorizer(dtype=<class 'numpy.float64'>, separator='=', sort=True,
               sparse=False)
```


Also what I meant by
>  We're not using ellipsis anymore

is that this PR won't do ellipsis. Of course the current implementation still does.

I like @amueller's idea: the default value of `print_changed_only` could be `'auto'`, which would use `True` iff the repr is more than x characters, else `False`.

Reply:
In CountVectorizer (but not DictVectorizer) you can pass a vocabulary in ``__init__``: https://scikit-learn.org/dev/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html

Reply:
```python
from sklearn.feature_extraction.text import CountVectorizer
vocabulary = {i: i for i in range(100)}
vectorizer = CountVectorizer(vocabulary=vocabulary)
vectorizer
```

```
CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None,
                vocabulary={0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7,
                            8: 8, 9: 9, 10: 10, 11: 11, 12: 12, 13: 13, 14: 14,
                            15: 15, 16: 16, 17: 17, 18: 18, 19: 19, 20: 20,
                            21: 21, 22: 22, 23: 23, 24: 24, 25: 25, 26: 26,
                            27: 27, 28: 28, 29: 29, 30: 30, 31: 31, 32: 32,
                            33: 33, 34: 34, 35: 35, 36: 36, 37: 37, 38: 38,
                            39: 39, 40: 40, 41: 41, 42: 42, 43: 43, 44: 44,
                            45: 45, 46: 46, 47: 47, 48: 48, 49: 49, 50: 50,
                            51: 51, 52: 52, 53: 53, 54: 54, 55: 55, 56: 56,
                            57: 57, 58: 58, 59: 59, 60: 60, 61: 61, 62: 62,
                            63: 63, 64: 64, 65: 65, 66: 66, 67: 67, 68: 68,
                            69: 69, 70: 70, 71: 71, 72: 72, 73: 73, 74: 74,
                            75: 75, 76: 76, 77: 77, 78: 78, 79: 79, 80: 80,
                            81: 81, 82: 82, 83: 83, 84: 84, 85: 85, 86: 86,
                            87: 87, 88: 88, 89: 89, 90: 90, 91: 91, 92: 92,
                            93: 93, 94: 94, 95: 95, 96: 96, 97: 97, 98: 98,
                            99: 99})
```

The whole dict is printed even if we go up to `1000` instead of `100`. I'm afraid the `print_only_changed` parameter won't be of any help here :/

Reply:
Looks like we do need "...". Maybe only for non-estimators though?

Reply:
And this current test is considering the case of a non-estimator parameter, so it should be more or less maintained, even if the number of characters differs.

Reply:
Got it, I'll try to add ellipsis for non-estimators soon

Reply:
Well... I looked into it and I think that adding ellipsis for non-estimator objects is going to be unreasonably difficult. Maybe even impossible.

The main reason is that, e.g. for a dict, `pprint` never returns a string representing this object. It will instead add '{' to a stream (typically stdout), then it will add all the key-values representations chunk by chunk, and then '}'. 

So there is no way to do something like

```
if len(my_dict_repr) > limit:
    # add ellipsis
```

because `my_dict_repr` does not exist.

We could easily add the ellipsis **after** the call to `pprint`, but would likely "cut" objects in arbitrary places.

Reply:
how about we call pprint to create the repr, then check if it's too long, and if it's too long we throw it out and create a new one using more and more aggressive heuristics that make it shorter.
The first one is ``changed_only=True``, the second one could be additionally replacing all non-scalar values by ellipsis and the fallback is Just print the estimator name with ellipsis ``CountVectorizer(...)```.

Reply:
I think IPython uses its own [pprint](https://ipython.readthedocs.io/en/stable/api/generated/IPython.lib.pretty.html) module.

We could count the number of characters returned by `write` but that would require changing every single call to `write`, so we'd need to copy/paste the whole `pprint` module.

Even then, how do we alter the stream afterwards? As far as I can tell, going back in a `StringIO` object with `seek` is impossible:

```
seek(pos, whence=0, /) method of _io.StringIO instance
    Change stream position.
    
    Seek to character offset pos relative to position indicated by whence:
        0  Start of stream (the default).  pos should be >= 0;
        1  Current position - pos must be 0;
        2  End of stream - pos must be 0.
    Returns the new absolute position.
```

Reply:
How about this:

The number of non-estimator parameters that are potentially very long (such as the `vocabulary` param of `CountVectorizer`) is pretty small, right?

So instead of trying to solve the problem in `PrettyPrinter`, we could try to solve it upstream, namely in a custom `get_param` function where we would replace those potentially long parameters by a shortened version. In the case of `vocabulary` that could be, for example, keeping only a subset of key-values, or replacing it by an empty dict. Or by a 'custom_vocabulary' str object.

It's really a case-by-case basis but I don't think we would have a lot of such cases?

This could be an additional step to @amueller's proposed process above:
- try pprint
- try again with those potentially long parameters removed
- try again with changed_only=True (and without the long params)
- ...

Reply:
Here are a few examples with @amueller's idea of using ellipsis on non-basic types (lists, dicts, tuples, etc.)

It's pretty easy to implement, I can push changes if it's satisfactory for you. This fixes the `CountVectorizer/vocabulary` issue but `Pipeline` and `GridSearch` aren't great.

<details>

```
Pipeline(memory=None,
         steps=...)

RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=RFE(estimator=LogisticRegression(C=1.0,
                                                                                                                     class_weight=None,
                                                                                                                     dual=False,
                                                                                                                     fit_intercept=True,
                                                                                                                     intercept_scaling=1,
                                                                                                                     max_iter=100,
                                                                                                                     multi_class='warn',
                                                                                                                     n_jobs=None,
                                                                                                                     penalty='l2',
                                                                                                                     random_state=None,
                                                                                                                     solver='warn',
                                                                                                                     tol=0.0001,
                                                                                                                     verbose=0,
                                                                                                                     warm_start=False),
                                                                                        n_features_to_select=None,
                                                                                        step=1,
                                                                                        verbose=0),
                                                                          n_features_to_select=None,
                                                                          step=1,
                                                                          verbose=0),
                                                            n_features_to_select=None,
                                                            step=1, verbose=0),
                                              n_features_to_select=None, step=1,
                                              verbose=0),
                                n_features_to_select=None, step=1, verbose=0),
                  n_features_to_select=None, step=1, verbose=0),
    n_features_to_select=None, step=1, verbose=0)


GridSearchCV(cv=5, error_score='raise-deprecating',
             estimator=SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
                           decision_function_shape='ovr', degree=3,
                           gamma='auto_deprecated', kernel='rbf', max_iter=-1,
                           probability=False, random_state=None, shrinking=True,
                           tol=0.001, verbose=False),
             iid='warn', n_jobs=None,
             param_grid=...,
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)


GridSearchCV(cv=3, error_score='raise-deprecating',
             ('estimator', Pipeline(memory=None,
         steps=...)),
             iid='warn', n_jobs=1,
             param_grid=...,
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None,
                vocabulary=...)

```
</details>

Reply:
I would always try the changed_only first and if that's not short enough, I'd do changed_only plus this heuristic. So these would look quite different, I think?

Reply:
They would look slightly shorter yes. But we would still be using ellipsis on the `steps` or `param_grid` parameters which IMO makes the representations not very informative.

Reply:
yeah but if they are super long otherwise they won't be informative anyway, right?

Reply:
Well in the case of `CountVectorizer/vocabulary` for example, what may make the `__repr__` super long is likely to be  `vocabulary`, not the `steps` or `param_grid` parameters.

Ideally, we should use ellipsis on `vocabulary` only. If it's still too long they we could use ellipsis more. But that would require hardcoding which parameters should be "ellipsed" first.

Reply:
I don't think we need to hardcode which parameters should be ellipsed. If the vocabulary has more than `MAX_ELEMENTS_TO_SHOW` entries then we show a few and elide the rest.

Reply:
@amueller , @jnothman ,

As of 5e23560 I have added:

- Ellipsis based on the number of elements in lists, tuple, dicts
- `__repr__` now tries in this order:
  - the original representation
  - repr with ellipsis on sequences with a lot of elements
  - repr with ellipsis on sequences with a lot of elements + changed_only = True
  - repr with ellipsis on sequences with a lot of elements + changed_only = True + bruteforce ellipsis that cuts the repr right in the middle.
- tests for all these features

Hope this will be it ^^ 🤞🤞🤞🤞

Reply:
I don't think it adds a lot of extra complexity, it's just a few lines in `__repr__`. We've had the `changed_only` logic right from the beginning of this PR, now we just force it if the repr is too long.

Reply:
I think it's a nice place to show-case that functionality and it doesn't look that complex to me tbh. You mean complex in terms of code or complex for the user to understand?

Reply:
I mean unnecessary and unjustified and hence complex by all measures. But I should check the code to make sure my understanding is correct